pkgs/build-support/dotnet/build-dotnet-module/default.nix

```diff
@@ -13,210 +13,221 @@
-  # The actual version attr is saved in InformationalVersion, which accepts an arbitrary string
-  versionForDotnet = if !(lib.hasAttr "version" args) || args.version == null
-  then null else let
-    components = lib.pipe args.version [
-      lib.splitVersion
-      (lib.filter (x: (lib.strings.match "[0-9]+" x) != null))
-      (lib.filter (x: (lib.toIntBase10 x) < 65535)) # one version component in dotnet has to fit in 16 bits
-    ];
-  in if (lib.length components) == 0
-  then null
-  else lib.concatStringsSep "." ((lib.take 4 components)
-    ++ (if (lib.length components) < 4
-       then lib.replicate (4 - (lib.length components)) "0"
-       else [ ]));
-
-  makeWrapperArgs = args.makeWrapperArgs or [ ] ++ [
-    "--prefix" "LD_LIBRARY_PATH" ":" "${dotnet-sdk.icu}/lib"
-  ];
-
-  # Stripping breaks the executable
-  dontStrip = args.dontStrip or true;
-
-  # gappsWrapperArgs gets included when wrapping for dotnet, as to avoid double wrapping
-  dontWrapGApps = args.dontWrapGApps or true;
-
-  # propagate the runtime sandbox profile since the contents apply to published
-  # executables
-  propagatedSandboxProfile = toString dotnet-runtime.__propagatedSandboxProfile;
-
-  passthru = {
-    nugetDeps = _nugetDeps;
-  } // lib.optionalAttrs (!lib.isDerivation nugetDeps) {
-    fetch-deps = let
-      pkg = finalAttrs.finalPackage.overrideAttrs (old: {
-        buildInputs = lib.remove _nugetDeps old.buildInputs;
-        keepNugetConfig = true;
-      } // lib.optionalAttrs (runtimeId == null) {
-        dotnetRuntimeIds = map (system: systemToDotnetRid system) platforms;
-      });
-
-      drv = builtins.unsafeDiscardOutputDependency pkg.drvPath;
-
-      innerScript = substituteAll {
-        src = ./fetch-deps.sh;
-        isExecutable = true;
-        defaultDepsFile =
-          # Wire in the nugetDeps file such that running the script with no args
-          # runs it agains the correct deps file by default.
-          # Note that toString is necessary here as it results in the path at
-          # eval time (i.e. to the file in your local Nixpkgs checkout) rather
-          # than the Nix store path of the path after it's been imported.
-          if lib.isPath nugetDepsFile
-          && !lib.hasPrefix "${builtins.storeDir}/" (toString nugetDepsFile) then
-            toString nugetDepsFile
-          else
-            ''$(mktemp -t "${pname}-deps-XXXXXX.nix")'';
-        nugetToNix = (nuget-to-nix.override { inherit dotnet-sdk; });
-      };
-
-      in writeShellScript "${name}-fetch-deps" ''
-        NIX_BUILD_SHELL="${runtimeShell}" exec ${nix}/bin/nix-shell \
-          --pure --run 'source "${innerScript}"' "${drv}"
-      '';
-  } // args.passthru or { };
-
-  meta = (args.meta or { }) // { inherit platforms; };
-})
+  fnOrAttrs:
+    stdenvNoCC.mkDerivation (finalAttrs:
+      let
+        _args =
+          if builtins.isFunction fnOrAttrs
+          then fnOrAttrs finalAttrs
+          else fnOrAttrs;
```

Comment:
Would it make sense for the `finalAttrs` passed to the supplied fn to be different from the `finalAttrs` from the underlying `mkDerivation`?

To put this another way, is it confusing for the dotnet-specific arguments to _not_ be present in the `finalAttrs` seen by calling packages?

I wonder if this should be doing something like `fnOrAttrs (_args // finalAttrs)`, although I'm not sure that would be lazy enough to avoid infinite recursion...

Reply:
Are you thinking of `removeAttrs _args [ "nugetDeps" ]`?  As an aside: lots of other things _should_ be in that list, or passed in a separate attribute list or something.

Reply:
Sorry, I should probably be asleep so I'm struggling to make myself clear! :zzz: 

I mean that package maintainers would probably expect attrs that are passed to `buildDotnetModule` but not known to `mkDerivation` to be present in the `finalAttrs` they see in their package. Even if they're not present in the _underlying_ `finalAttrs` provided by `mkDerivation`?

I guess if we're only removing `"nugetDeps"` and if `mkDerivation` includes any arbitrary/unknown attrs in `finalAttrs` then this may be a non-issue currently though?

Reply:
That's what I thought you meant.  I was just double-checking because I think it really is only that one attr, and everything else makes it to stdenv (currently, but probably shouldn't).

I _think_ `_args // finalAttrs` should be fine, and it feels more intuitive that way.  I've been looking for anything similar in other language infrastructure, but no luck so far.  Python just has:

```
  # Derivations built with `buildPythonPackage` can already be overridden with `override`, `overrideAttrs`, and `overrideDerivation`.
  # This function introduces `overridePythonAttrs` and it overrides the call to `buildPythonPackage`.
  makeOverridablePythonPackage = f: lib.mirrorFunctionArgs f (origArgs:
```

But overriding (`overrideDotnetAttrs`?) is a whole other can of worms.